adapter/src/hermes/adapter/stdio/stdio.cc

```diff
@@ -0,0 +1,993 @@
+        stat.st_bkid =
+            std::make_shared<hapi::Bucket>(path_str, mdm->GetHermes(), ctx);
+        mdm->Create(ret, stat);
+      } else {
+        // TODO(hari): @error_handling invalid fh.
+        ret = nullptr;
+      }
+    } else {
+      LOG(INFO) << "File opened before by adapter" << std::endl;
+      existing.first.ref_count++;
+      struct timespec ts;
+      timespec_get(&ts, TIME_UTC);
+      existing.first.st_atim = ts;
+      existing.first.st_ctim = ts;
+      mdm->Update(ret, existing.first);
+    }
+  }
+  return ret;
+}
+
+FILE *open_internal(const std::string &path_str, const char *mode) {
+  FILE *ret;
+  MAP_OR_FAIL(fopen);
+  ret = real_fopen_(path_str.c_str(), mode);
+  ret = simple_open(ret, path_str.c_str(), mode);
+  return ret;
+}
+
+FILE *reopen_internal(const std::string &path_str, const char *mode,
+                      FILE *stream) {
+  FILE *ret;
+  MAP_OR_FAIL(freopen);
+  auto mdm = hermes::adapter::Singleton<MetadataManager>::GetInstance();
+  ret = real_freopen_(path_str.c_str(), mode, stream);
+  if (!ret) {
+    return ret;
+  } else {
+    auto existing = mdm->Find(ret);
+    if (!existing.second) {
+      LOG(INFO) << "Reopen file for filename " << path_str << " in mode "
+                << mode << std::endl;
+      LOG(INFO) << "File not opened before by adapter" << std::endl;
+      return nullptr;
+    } else {
+      LOG(INFO) << "File opened before by adapter" << std::endl;
+      struct timespec ts;
+      timespec_get(&ts, TIME_UTC);
+      existing.first.st_atim = ts;
+      existing.first.st_ctim = ts;
+      mdm->Update(ret, existing.first);
+    }
+  }
+  return ret;
+}
+
+size_t write_internal(std::pair<AdapterStat, bool> &existing, const void *ptr,
+                      size_t total_size, FILE *fp) {
+  LOG(INFO) << "Write called for filename: "
+            << existing.first.st_bkid->GetName()
+            << " on offset: " << existing.first.st_ptr
+            << " and size: " << total_size << std::endl;
+  size_t ret;
+  auto mdm = hermes::adapter::Singleton<MetadataManager>::GetInstance();
+  auto mapper = MapperFactory().Get(kMapperType);
+  auto mapping = mapper->map(
+      FileStruct(mdm->Convert(fp), existing.first.st_ptr, total_size));
+  size_t data_offset = 0;
+  auto filename = existing.first.st_bkid->GetName();
+  LOG(INFO) << "Mapping for read has " << mapping.size() << " mapping."
+            << std::endl;
+  for (const auto &item : mapping) {
+    hapi::Context ctx;
+    auto blob_exists =
+        existing.first.st_bkid->ContainsBlob(item.second.blob_name_);
+    hapi::Blob put_data((unsigned char *)ptr + data_offset,
+                        (unsigned char *)ptr + data_offset + item.first.size_);
+    existing.first.st_blobs.emplace(item.second.blob_name_);
+    if (!blob_exists || item.second.size_ == kPageSize) {
+      LOG(INFO) << "Overwrite blob " << item.second.blob_name_
+                << " of size:" << item.second.size_ << "." << std::endl;
```

Comment:
We're not overwriting the blob if it doesn't exist, right?

Reply:
It has both. Create or Overrite. I updated the comment.